File: src/main/java/com/github/albertosh/flitetrakr/reactive/usecases/connectionsbelowprice/ConnectionsBelowPriceUseCase.java
package com.github.albertosh.flitetrakr.reactive.usecases.connectionsbelowprice;

import com.github.albertosh.flitetrakr.model.Connection;
import com.github.albertosh.flitetrakr.model.MultipleConnections;
import com.github.albertosh.flitetrakr.reactive.services.IConnectionService;
import com.github.albertosh.flitetrakr.reactive.usecases.cheapestconnection.CheapestConnectionUseCaseError;
import com.github.albertosh.flitetrakr.reactive.usecases.cheapestconnection.CheapestConnectionUseCaseInput;
import com.github.albertosh.flitetrakr.reactive.usecases.cheapestconnection.ICheapestConnectionUseCase;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import rx.Observable;
import rx.Single;
import rx.functions.Func1;
import rx.functions.Func2;
import rx.schedulers.Schedulers;

public class ConnectionsBelowPriceUseCase
        implements IConnectionsBelowPriceUseCase {

    private final ICheapestConnectionUseCase cheapestConnectionUseCase;
    private final IConnectionService connectionService;

    public ConnectionsBelowPriceUseCase(ICheapestConnectionUseCase cheapestConnectionUseCase, IConnectionService connectionService) {
        this.cheapestConnectionUseCase = cheapestConnectionUseCase;
        this.connectionService = connectionService;
    }

    @Override
    public Observable<List<MultipleConnections>> execute(ConnectionsBelowPriceUseCaseInput input) {
        return calculateCheapestConnection(input)
                .toObservable()
                .onErrorResumeNext(error -> {
                    if (error instanceof CheapestConnectionUseCaseError) {
                        CheapestConnectionUseCaseError cheapestConnectionUseCaseError = (CheapestConnectionUseCaseError) error;
                        if (cheapestConnectionUseCaseError.getErrorType()
                                .equals(CheapestConnectionUseCaseError.ErrorType.CONNECTION_NOT_FOUND)) {
                            return Observable.error(ConnectionsBelowPriceUseCaseError.connectionNotFound());
                        }
                    }
                    return Observable.error(error);
                })
                .flatMap(cheapestConnection -> {
                    if (cheapestConnection > input.getPrice()) {
                        // The cheapest connection is more expensive than our output
                        // We're done...
                        return Observable.just(Collections.emptyList());
                    } else {
                        Observable<MultipleConnections> output = performDPSSearch(input);

                        return output
                                .scan(Collections.emptyList(), new Func2<List<MultipleConnections>, MultipleConnections, List<MultipleConnections>>() {
                                    @Override
                                    public List<MultipleConnections> call(List<MultipleConnections> accum, MultipleConnections newValue) {
                                        List<MultipleConnections> result = new ArrayList<>(accum);
                                        result.add(newValue);
                                        Collections.sort(result);
                                        return result;
                                    }
                                })
                                .skip(1); // Ignore first empty list
                    }
                });
    }

    private Single<Integer> calculateCheapestConnection(ConnectionsBelowPriceUseCaseInput input) {
        CheapestConnectionUseCaseInput cheapestConnectionUseCaseInput = new CheapestConnectionUseCaseInput.Builder()
                .from(input.getFrom())
                .to(input.getTo())
                .build();
        return cheapestConnectionUseCase.execute(cheapestConnectionUseCaseInput)
                .map(output -> output.getPrice());
    }

    private Observable<MultipleConnections> performDPSSearch(ConnectionsBelowPriceUseCaseInput input) {

        MultipleConnections.Builder connectionBuilder =
                new MultipleConnections.Builder(input.getFrom());

        Observable<MultipleConnections> firstItem = Observable.empty();
        // Concrete case of starting and ending at the same place
        if (input.getFrom().equals(input.getTo())) {
            firstItem = Observable.just(
                    new MultipleConnections.Builder(input.getFrom()).build());
        }

        return Observable.concat(firstItem, performDPSSearch(input.getFrom(), input.getTo(), input.getPrice(), connectionBuilder));
    }

    private Observable<MultipleConnections> performDPSSearch(String from, String to, Integer price,
                                                             MultipleConnections.Builder connectionBuilder) {

        return connectionService.getDestiniesFromCity(from)
                .flatMap(city ->
                        connectionService.recoverConnection(from, city)
                                .toObservable()
                                .filter(new Func1<Connection, Boolean>() {
                                    @Override
                                    public Boolean call(Connection connection) {
                                        return (connection.getPrice() <= price);
                                    }
                                })
                                .flatMap(connection -> {
                                    MultipleConnections.Builder newBuilder = connectionBuilder.clone();
                                    newBuilder.plusPrice(connection.getPrice());
                                    newBuilder.withCity(city);

                                    if (city.equals(to))
                                        return Observable.just(newBuilder.build())
                                                .concatWith(performDPSSearch(city, to,
                                                        price - connection.getPrice(), newBuilder));
                                    else
                                        return performDPSSearch(city, to,
                                                price - connection.getPrice(), newBuilder);
                                })
                                .subscribeOn(Schedulers.computation())
                );
    }


}

File: flitetrakr_reactive_kotlin/src/main/kotlin/com/github/albertosh/flitetrakr/usecases/ConnectionsBelowPriceUseCase.kt
package com.github.albertosh.flitetrakr.usecases

import com.github.albertosh.flitetrakr.model.MultipleConnections
import com.github.albertosh.flitetrakr.services.IConnectionService
import com.github.albertosh.flitetrakr.util.language.LanguageUtils
import com.github.albertosh.flitetrakr.util.language.Message
import io.reactivex.Flowable
import io.reactivex.SingleSource
import org.reactivestreams.Publisher

interface IConnectionsBelowPriceUseCase {

    fun execute(input: ConnectionsBelowPriceUseCaseInput): Flowable<ConnectionsBelowPriceUseCaseOutput>

}

data class ConnectionsBelowPriceUseCaseInput(val from: String, val to: String, val price: Int)

data class ConnectionsBelowPriceUseCaseOutput(val connections: List<MultipleConnections>)

sealed class ConnectionsBelowPriceUseCaseError(message: String) : RuntimeException(message) {
    object connectionNotFound : ConnectionsBelowPriceUseCaseError(
            LanguageUtils.getMessage(Message.CONNECTION_NOT_FOUND))
}


class ConnectionsBelowPriceUseCase(
        private val service: IConnectionService,
        private val cheapestConnectionUseCase: ICheapestConnectionUseCase)
: IConnectionsBelowPriceUseCase {

    override fun execute(input: ConnectionsBelowPriceUseCaseInput): Flowable<ConnectionsBelowPriceUseCaseOutput> =
            calculateCheapestConnection(input)
                    .onErrorResumeNext { error ->
                        when (error) {
                            is CheapestConnectionUseCaseError.connectionNotFound -> SingleSource {
                                it.onError(ConnectionsBelowPriceUseCaseError.connectionNotFound)
                            }
                            else -> SingleSource {
                                it.onError(error)
                            }
                        }
                    }
                    .toFlowable()
                    .flatMap {
                        if (it > input.price) {
                            Publisher<ConnectionsBelowPriceUseCaseOutput> {
                                it.onNext(ConnectionsBelowPriceUseCaseOutput(emptyList()))
                                it.onComplete()
                            }
                        } else {
                            performDPSSearch(input)
                                    .scan(emptyList<MultipleConnections>(), { accum, newValue ->
                                        accum.plus(newValue).sorted()
                                    })
                                    .skip(1) // Ignore first empty list
                                    .map { ConnectionsBelowPriceUseCaseOutput(it) }
                        }
                    }


    private fun calculateCheapestConnection(input: ConnectionsBelowPriceUseCaseInput) =
            cheapestConnectionUseCase.execute(
                    CheapestConnectionUseCaseInput(
                            from = input.from,
                            to = input.to
                    ))
                    .map { it.price }


    private fun performDPSSearch(input: ConnectionsBelowPriceUseCaseInput): Flowable<MultipleConnections> {
        val connections = MultipleConnections(mutableListOf(input.from), 0)
        val firstItem = if (input.from == input.to)
            Flowable.just(connections)
        else
            Flowable.empty()
        return Flowable.concat(firstItem, performDPSSearch(input.from, input.to, input.price, connections))
    }


    private fun performDPSSearch(from: String, to: String, price: Int,
                                 connections: MultipleConnections): Flowable<MultipleConnections> =
            service.getDestiniesFromCity(from)
                    .flatMap { city ->
                        service.recoverConnection(from, city)
                                .filter { it.price <= price }
                                .toFlowable()
                                .flatMap {
                                    val newConnections = connections.copy(
                                            cities = connections.cities.plus(city),
                                            price = connections.price + it.price
                                    )

                                    if (city == to)
                                        Flowable.concat(
                                                Flowable.just(newConnections),
                                                performDPSSearch(city, to,
                                                        price - it.price, newConnections)
                                        )
                                    else
                                        performDPSSearch(city, to,
                                                price - it.price, newConnections)

                                }

                    }

}
